pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1309,3 +1343,100 @@ func initializeAnnotations(cd *hivev1.ClusterDeployment) {
 		cd.Annotations = map[string]string{}
 	}
 }
+
+// mergePullSecrets merges the global pull secret JSON (if defined) with the cluster's pull secret JSON (if defined)
+// An error will be returned if neither is defined
+func (r *ReconcileClusterDeployment) mergePullSecrets(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (string, error) {
+	var localPullSecret string
+	var err error
+
+	// For code readability let's call the pull secret in cluster deployment config as local pull secret
+	if cd.Spec.PullSecret != nil {
+		localPullSecret, err = controllerutils.LoadSecretData(r.Client, cd.Spec.PullSecret.Name, cd.Namespace, corev1.DockerConfigJsonKey)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return "", err
+			}
+		}
+	}
+
+	// Check if global pull secret from env as it comes from hive config
+	globalPullSecret := os.Getenv("GLOBAL_PULL_SECRET")
+
+	switch {
+	case globalPullSecret != "" && localPullSecret != "":
+		// Merge local pullSecret and globalPullSecret. If both pull secrets have same registry name
+		// then the merged pull secret will have registry secret from local pull secret
+		pullSecret, err := controllerutils.MergeJsons(globalPullSecret, localPullSecret, cdLog)
+		if err != nil {
+			errMsg := "unable to merge global pull secret with local pull secret"
+			cdLog.WithError(err).Error(errMsg)
+			return "", errors.Wrap(err, errMsg)
+		}
+		return pullSecret, nil
+	case globalPullSecret != "":
+		return globalPullSecret, nil
+	case localPullSecret != "":
+		return localPullSecret, nil
+	default:
+		errMsg := "clusterdeployment must specify pull secret since hiveconfig does not specify a global pull secret"
+		cdLog.Error(errMsg)
+		return "", errors.New(errMsg)
+	}
+}
+
+// updatePullSecretInfo adds pull secret information in cluster deployment and cluster deployment status.
+// It returns true when cluster deployment status has been updated.
+func (r *ReconcileClusterDeployment) updatePullSecretInfo(pullSecret string, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (bool, error) {
+	var err error
+	pullSecretObjExists := true
+	existingPullSecretObj := &corev1.Secret{}
+	mergedSecretName := constants.GetMergedPullSecretName(cd)
+	err = r.Get(context.TODO(), types.NamespacedName{Name: mergedSecretName, Namespace: cd.Namespace}, existingPullSecretObj)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			cdLog.Info("Existing pull secret object not found")
+			pullSecretObjExists = false
+		} else {
+			return false, errors.Wrap(err, "Error getting pull secret from cluster deployment")
+		}
+	}
+
+	if pullSecretObjExists {
+		existingPullSecret, ok := existingPullSecretObj.Data[corev1.DockerConfigJsonKey]
+		if !ok {
+			return false, errors.New(fmt.Sprintf("Pull secret %s did not contain key %s", mergedSecretName, corev1.DockerConfigJsonKey))
+		}
+		fmt.Println(string(existingPullSecret))
```

Comment:
Will need to remove this before we can merge, don't want to log any pull secrets.